File: 438-find-all-anagrams-in-a-string/solution.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        unordered_map<char, int> need, window;
        for (auto c: p) need[c] ++;
        int left = 0, right = 0, valid = 0;
        vector<int> res;

        while (right < s.size()) {
            char c = s[right];
            right ++;
            if (need.count(c)) {
                window[c] ++;
                if (need[c] == window[c]) {
                    valid ++;
                }
            }

            while (right - left >= p.size()) {
                if (valid == need.size()) {
                    res.push_back(left);
                }
                char d = s[left];
                left ++;
                if (need.count(d)) {
                    if (need[d] == window[d]) {
                        valid --;
                    }
                    window[d] --;
                }
            }
        }
        return res;
    }
};

File: 438-find-all-anagrams-in-a-string/solution.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        need = dict()
        for v in p:
            if v not in need:
                need[v] = 0
            need[v] += 1
        window = dict()

        left, right = 0, 0
        start, length = 0, sys.maxsize
        valid = 0

        res = []

        while (right < len(s)):
            c = s[right]
            if c in need:
                if c not in window:
                    window[c] = 0
                window[c] += 1
                if window[c] == need[c]:
                    valid += 1
            
            right += 1
            while (right - left >= len(p)):
                if valid == len(need.keys()):
                    res.append(left)
                d = s[left]
                left += 1
                if d in need:
                    if window[d] == need[d]:
                        valid -= 1
                    window[d] -= 1
        return res
